austation/code/modules/food_and_drinks/food/kitchen_machinery/cake_printer.dm

```diff
@@ -0,0 +1,190 @@
+	var/obj/item/item_scanned
+	var/processing = FALSE
+	var/efficiency = 1
+	var/speed = 1
+	var/max_fuel = 100
+	var/static/list/cake_blacklist = typecacheof(list(
+		/obj/item/screwdriver,
+		/obj/item/crowbar,
+		/obj/item/wrench,
+		/obj/item/wirecutters,
+		/obj/item/multitool,
+		/obj/item/weldingtool,
+		/obj/item/reagent_containers/glass,
+		/obj/item/reagent_containers/syringe,
+		/obj/item/reagent_containers/food/condiment))
+	var/fuel = 0
+	var/fuelcost = 0
+
+/obj/machinery/cake_printer/update_icon()
+	if(panel_open)
+		icon_state = "kek-printer-o"
+	else if(!src.processing)
+		icon_state = "kek-printer-stand"
+	else
+		icon_state = "kek-printer-work"
+	return
+
+/obj/machinery/cake_printer/Initialize()
+	. = ..()
+	component_parts = list()
+	component_parts += new /obj/item/circuitboard/machine/cake_printer(null)
+	component_parts += new /obj/item/stock_parts/manipulator(null)
+	component_parts += new /obj/item/stock_parts/matter_bin(null)
+	component_parts += new /obj/item/reagent_containers/glass/beaker/large(null)
+	RefreshParts()
+
+/obj/machinery/cake_printer/RefreshParts()
+	var/max_storage = 100
+	for(var/obj/item/stock_parts/manipulator/M in component_parts)
+		efficiency = M.rating
+	for(var/obj/item/stock_parts/manipulator/P in component_parts)
+		speed = P.rating
+	for(var/obj/item/stock_parts/matter_bin/B in component_parts)
+		max_storage = 100 * B.rating
+	max_fuel = max_storage
+
+/obj/machinery/cake_printer/emag_act(mob/user)
+	. = ..()
+	if(!(obj_flags & EMAGGED)) //If it is not already emagged, emag it.
+		to_chat(user, "<span class='warning'>You disable the [src]'s safety features, allowing it's cakes to be toxic.</span>")
+		do_sparks(5, TRUE, src)
+		obj_flags |= EMAGGED
+		log_game("[key_name(user)] emagged [src]")
+		message_admins("[key_name_admin(user)] emagged [src]")
+	else
+		to_chat(user, "<span class='warning'>The status display on [src] is already too damaged to short it again.</span>")
+		return
+
+/obj/machinery/cake_printer/attackby(obj/item/I, mob/user)
+	if(user.a_intent == INTENT_HARM)
+		return ..()
+	if(processing)
+		to_chat(user, "<span class='warning'>[src] is printing!</span>")
+		return
+	if(default_deconstruction_screwdriver(user, "kek-printer-o", "kek-printer-o" ,I))
+		update_icon()
+		return ..()
+	if(default_deconstruction_crowbar(I))
+		return ..()
+	if(panel_open)
+		to_chat(user, "<span class='warning'>Close the maintenance panel first.</span>")
+		return ..()
+	if(istype(I, /obj/item/reagent_containers))
+		if(I.reagents.has_reagent(/datum/reagent/consumable/synthetic_cake_batter))
+			var/obj/item/reagent_containers/reagent_source = I
+			if(fuel == max_fuel)
+				to_chat(user, "<span class='warning'>The machine's tank is full!</span>")
+				return ..()
+			if((10 * (reagent_source.amount_per_transfer_from_this) > max_fuel))
+				to_chat(user, "<span class='warning'>Too much reagent per transfer, the tank can't hold it all!</span>")
+				return ..()
```

Comment:
It'd be better to make it transfer the max amount it can until the machine is full, like most reagent transfers do. There's probably a proc for it already